torch/distributions/wishart.py

```diff
@@ -0,0 +1,198 @@
+import math
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import _standard_normal, lazy_property
+from torch.distributions.multivariate_normal import _precision_to_scale_tril
+
+
+def _mvdigamma(x, p) -> torch.Tensor:
+    assert x > (p - 1) / 2, "Wrong domain for digamma function."
+    return torch.digamma(x - torch.arange(p) / 2).sum()
+
+class Wishart(Distribution):
+    r"""
+    Creates a Wishart distribution parameterized by a square root matrix of symmetric matrix.
+    The Wishart distribution can be parameterized either in terms of
+    an outer product of general square root matrix e,g.,
+    :math:`\mathbf{\Sigma} = \mathbf{P}\mathbf{D}\mathbf{P}^\top = \mathbf{P'}\mathbf{P'}^\top` or
+    an outer product of cholesky decomposition :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top`
+    can be obtained via Cholesky decomposition of the covariance.
+    Example:
+        >>> m = Wishart(torch.eye(2), torch.Tensor([2]))
+        >>> m.sample()  #Wishart distributed with mean=`df * I` and
+                        #variance(x_ij)=`df` for i != j and variance(x_ij)=`2 * df` for i == j
+    Args:
+        covariance_matrix (Tensor): positive-definite covariance matrix
+        precision_matrix (Tensor): positive-definite precision matrix
+        scale_tril (Tensor): lower-triangular factor of covariance, with positive-valued diagonal
+        df (Tensor): real-valued parameter > larger than dimension of Square matrix
+    Note:
+        Only one of :attr:`covariance_matrix` or :attr:`precision_matrix` or
+        :attr:`scale_tril` can be specified.
+        Using :attr:`scale_tril` will be more efficient: all computations internally
+        are based on :attr:`scale_tril`. If :attr:`covariance_matrix` or
+        :attr:`precision_matrix` is passed instead, it is only used to compute
+        the corresponding lower triangular matrices using a Cholesky decomposition.
+        'torch.distributions.LKJCholesky' is a restricted Wishart distribution.[1]
+
+    **References**
+
+    [1] `On equivalence of the LKJ distribution and the restricted Wishart distribution`,
+    Zhenxun Wang, Yunan Wu, Haitao Chu.
+    """
+    arg_constraints = {
+        'covariance_matrix': constraints.positive_definite,
+        'precision_matrix': constraints.positive_definite,
+        'scale_tril': constraints.positive_definite,
+        'df': None,
+    }
+    support = constraints.positive_definite
+    has_rsample = True
+
+    def __init__(self, covariance_matrix=None, precision_matrix=None, scale_tril=None, df=None,
+                 bartlett_decomposition=True, validate_args=None):
```

Comment:
We should document the `bartlett_decomposition` argument too. When is it a good idea to set this to `False`?

Reply:
I guess this is the original paper: [Link](https://projecteuclid.org/journals/annals-of-mathematical-statistics/volume-30/issue-1/Bartlett-Decomposition-and-Wishart-Distribution/10.1214/aoms/1177706379.full)
I will add the description to the docs soon!

Reply:
Using Bartlett decomposition is always better for following aspects:
1. Flexibility: Origianal Sampling requires 'df' parameter to be positive interger. Bartlett decomposition requires positive real 'df'.
2. Computation: Original sampling requires to sample (dim x df) normal sample given positive interger 'df'. Bartlett decomposition.

Reply:
If that is so, then we can just remove this option and use it as a default?

Reply:
Sure! I will set it as default and document it.